docs/architecture/adr-d001-inter-validator-set-messaging.md

```diff
@@ -122,40 +125,53 @@ through another Validator) to any other Validator which is a member of the same
 ### What systems will be affected?
 
 * Tenderdash
-* ABCI App
+* ABCI App (Dash Drive)
 
 ### What new data structures are needed, what data structures need changes?
 
-1. Separate Validator Set member connection pool from Full node connection pool
-2. Additional tuning parameters:
-    1. Full node connection slots
-    2. Validator node connection slots
+Implemented `dashtypes.ValidatorAddress` which represents an address of a validator received from ABCI app.
 
 ### What new APIs will be needed, what APIs will change?
 
+#### ABCI Protocol
 In the ABCI protocol, we add a network address of each member of the active Validator Set to the
-`ResponseEndBlock.ValidatorSetUpdate.ValidatorUpdate` structure. The network address structure is as follows:
+`ResponseEndBlock.ValidatorSetUpdate.ValidatorUpdate` structure. The network address should be an URI:
+
+```uri
+tcp://<node_id>@<address>:<port>
+```
+
+where:
+
+* `<node_id>` - node ID (can be generated based on node P2P public key, see [p2p.PubKeyToID](https://github.com/dashevo/tenderdash/blob/20d1f91c0adb29d1b5d03a945b17513a368759e4/p2p/key.go#L45))
+* `<address>` - IP address of the validator node
+* `<port>` - p2p port number of the validator node
+
+The `ValidatorUpdate` structure looks as follows:
 
 ```protobuf
-enum NetAddressType {
-  IPv4 = 0;
-};
-
-message NetAddress {
-  reserved 1; // Reserved for compatibility with tendermint.p2p.NetAddress
-  reserved "id"; // Reserved for compatibility with tendermint.p2p.NetAddress
-  string address = 2;
-  uint32 port = 3;
-  NetAddressType address_type = 4;
+message ValidatorUpdate {
+  tendermint.crypto.PublicKey pub_key      = 1 [(gogoproto.nullable) = true];
+  int64                       power        = 2;
+  bytes                       pro_tx_hash  = 3;
+  string                      node_address = 4 [(gogoproto.nullable) = true];  // address of the Validator
 }
 ```
 
+
+If ABCI app cannot determine `<node_id>`, it can leave it empty. In this case, Tenderdash will try to retrieve a public
+key from remote host and derive node ID from it.
+
+#### P2P Handshake
+
+To retrieve node ID from a **remote node**, **local node** can establish a short-lived connection to the **remote node** to download its public key. This feature does not require any changes to the P2P protocol; it just uses existing capabilities.
+
+Public key request starts with a normal flow Diffie-Hellman protocol, as implemented in `conn.MakeSecretConnection()`. Once epheremal keys are exchanged between nodes, **local node** proceeds to  `shareAuthSignature()`. However, instead of sending its own `AuthSigMessage`, it waits for **remote node**'s auth signature to arrive.As the communication is asynchronous, the **remote node** sends its `AuthSigMessage`. That's when the **local node** closes the connection.
```

Comment:
Are we going to change it to double handshake as @QuantumExplorer proposed?

Reply:
Added section "Future work" with this information, also added to Security section.